src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.cs

```diff
@@ -141,23 +141,31 @@ public AssetProvider CreateAssetProvider(Checksum solutionChecksum, SolutionAsse
 
             async ValueTask<(Solution? solution, T result)> TryFastGetSolutionAndRunAsync()
             {
-                Solution solution;
+                // Ensure we keep this lazy-solution around (and alive in _solutionChecksumToLazySolution as well) as
+                // long as we're processing.
+                ReferenceCountedDisposable<LazySolution>? refCountedLazySolution;
                 using (await _gate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
                 {
                     if (_lastRequestedPrimaryBranchSolution.checksum == solutionChecksum)
                     {
-                        solution = _lastRequestedPrimaryBranchSolution.solution;
+                        refCountedLazySolution = _lastRequestedPrimaryBranchSolution.refCountedSolution.TryAddReference();
                     }
                     else if (_lastRequestedAnyBranchSolution.checksum == solutionChecksum)
                     {
-                        solution = _lastRequestedAnyBranchSolution.solution;
+                        refCountedLazySolution = _lastRequestedAnyBranchSolution.refCountedSolution.TryAddReference();
```

Comment:
talked offline - decided that contracts here that TryAddReference succeeded make sense because the existence of a checksum and refcountedsolution pair in the field means that there is at least 1 ref count to it